core/src/main/scala/chisel3/package.scala

```diff
@@ -210,29 +211,116 @@ package object chisel3 {
+      val t = sc.parts.map(_.replaceAll("%","%%"))
+      StringContext(t : _*).cf(args : _*)
+    }
+
+    /*
+       Custom string interpolator for generating formatted Printables : cf"..."
+     * High level algorithm is as follows
+     * Extending the algorithm mentioned in the regular scala "f" interpolator - except this has to be enhanced for handling Chisel Data types and Printables.
+     * /** The formatted string interpolator.
+     *  The `f` method works by assembling a format string from all the `parts` strings and using
+     *  `java.lang.String.format` to format all arguments with that format string. The format string is
+     *  obtained by concatenating all `parts` strings, and performing two transformations:
+     *   1. Let a _formatting position_ be a start of any `parts` string except the first one.
+     *      If a formatting position does not refer to a `%` character (which is assumed to
+     *      start a format specifier), then the string format specifier `%s` is inserted. This means any string with no format specifiers should behave same as p interpolator.
+     *   2. If format specifiers specifically target Chisel types - Like Data/Bits -  specific internal methods are called on that argument (like FirrtlFormat)
+     *   3. If format specifers target regular scala data types - java.lang.String.format method is called with specifier and argument to get the formatted string.
+     *
+     * Yet to document all the corener cases which this code tries to cover - but here are few of them
+     * 1. Allow calling %f or %2.2f (or something similar) on Int type data (Int, Short, Long) - this requires doing an explicit cast of the data before calling String.format
+     * 2. For bits - if no format specifier given (%s i.e) call .toPrintable to pick the default one. For regular data - let the String.format handle canonical case of %s too.
+     * Still testing this code - I am pretty sure this can be refactored and made more efficient / readable. WIP.
+     *
+     * /
+     */
+     */
+    def cf(args: Any*): Printable = {
+
+      // Handle literal % 
+      // Takes the part string - 
+      // - this is assumed to not have any format specifiers - already handled / removed before calling this function. 
+      // Only thing present is literal % if any which should ideally be with %%. 
+      // If not - then flag an error. 
+      // Return seq of Printables (either PString or Percent or both - nothing else
+      def percentSplitter(s : String): Seq[Printable] = {
+          if(s.isEmpty) return Seq(PString(""))
+	        val pieces = s.split("%%").toList.flatMap { p =>
+	          if (p.contains('%')) throw new UnknownFormatConversionException("Un-escaped % found")
+	          // Wrap in PString and intersperse the escaped percentages
+	          Seq(Percent, PString(p))
+	        }
+	        if(pieces.isEmpty) Seq(Percent) else pieces.tail  // Don't forget to drop the extra percent we put at the beginning
```

Comment:
```suggestion
          if (s.isEmpty) Seq(PString(""))
          else {
	    val pieces = s.split("%%").toList.flatMap { p =>
	      if (p.contains('%')) throw new UnknownFormatConversionException("Un-escaped % found")
	      // Wrap in PString and intersperse the escaped percentages
	      Seq(Percent, PString(p))
	    }
	    if (pieces.isEmpty) Seq(Percent) else pieces.tail  // Don't forget to drop the extra percent we put at the beginning
	  }
```
Some minor formatting (space between `if` and `(`) -- this would probably be done by scalafmt

Also `return` is frowned up on Scala so tweaked this to use `if-else`